src/Control/Middleware/RewriteHashLinksMiddleware.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+namespace SilverStripe\Control\Middleware;
+
+use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Control\Middleware\HTTPMiddleware;
+use SilverStripe\Core\Convert;
+use SilverStripe\Core\Config\Configurable;
+
+class RewriteHashLinksMiddleware implements HTTPMiddleware
+{
+    use Configurable;
+
+    /**
+     * Mime types to be handled by this middleware
+     * @var string
+     * @config SilverStripe\Control\Middleware\RewriteHashLinksMiddleware.handled_mime_types
+     */
+    private static $handled_mime_types = [
+        'text/html',
+        'application/xhtml+xml',
+    ];
+
+
+    /**
+     * Rewrites hash links in html responses
+     * @param HTTPRequest $request
+     * @param callable $delegate
+     * @return \SilverStripe\Control\HTTPResponse
+     */
+    public function process(HTTPRequest $request, callable $delegate)
+    {
+        /** @var \SilverStripe\Control\HTTPResponse $response **/
+        $response = $delegate($request);
+
+
+        $contentType = explode(';', $response->getHeader('content-type'));
+        $contentType = strtolower(trim(array_shift($contentType)));
+        if (in_array($contentType, $this->config()->handled_mime_types)) {
+            $body = $response->getBody();
+
+            if (strpos($body, '<base') !== false) {
+                $body = preg_replace('/(<a[^>]+href *= *)"#/i', '\\1"' . Convert::raw2att(preg_replace("/^(\\/)+/", "/", $_SERVER['REQUEST_URI'])) . '#', $body);
```

Comment:
Need to test for href='' singlequotes as well as href="" doublequotes

Reply:
`preg_replace("/^(\\/)+/"`

Probably cleaner to use str_replace here, I'm not entirely sure what characters are being replace due to the escaping

I'd also move `Convert::raw2att...` to the line above and assign it to a variable to increase readability

Reply:
I actually lifted that from SSViewer ;) seems to be replacing multiple back to back backslashes i.e `//` or `///` for example.

Reply:
I've implemented your other suggestions though here :)